source/lac/petsc_communication_pattern.cc

```diff
@@ -0,0 +1,516 @@
+      }                                                  \
+    while (0)
+
+namespace PETScWrappers
+{
+  CommunicationPattern::CommunicationPattern()
+    : sf(nullptr)
+  {}
+
+  CommunicationPattern::~CommunicationPattern()
+  {
+    clear();
+  }
+
+  void
+  CommunicationPattern::reinit(const types::global_dof_index local_size,
+                               const IndexSet &              ghost_indices,
+                               const MPI_Comm &              communicator)
+  {
+    clear();
+
+    PetscLayout layout;
+    AssertPETSc(PetscLayoutCreate(communicator, &layout));
+    AssertPETSc(PetscLayoutSetLocalSize(layout, local_size));
+    AssertPETSc(PetscLayoutSetUp(layout));
+
+    PetscInt start, end;
+    AssertPETSc(PetscLayoutGetRange(layout, &start, &end));
+
+    IndexSet want;
+    want.add_range(start, end);
+    want.add_indices(ghost_indices);
+    want.compress();
+
+    const PetscInt *idxs;
+    PetscInt        n;
+    IS              is = want.make_petsc_is(communicator);
+    AssertPETSc(ISGetLocalSize(is, &n));
+    AssertPETSc(ISGetIndices(is, &idxs));
+
+    AssertPETSc(PetscSFCreate(communicator, &sf));
+    AssertPETSc(
+      PetscSFSetGraphLayout(sf, layout, n, nullptr, PETSC_OWN_POINTER, idxs));
+    AssertPETSc(PetscSFSetUp(sf));
+
+    AssertPETSc(ISRestoreIndices(is, &idxs));
+    AssertPETSc(ISDestroy(&is));
+    AssertPETSc(PetscLayoutDestroy(&layout));
+  }
+
+  void
+  CommunicationPattern::reinit(const IndexSet &vector_space_vector_index_set,
+                               const IndexSet &read_write_vector_index_set,
+                               const MPI_Comm &communicator)
+  {
+    std::vector<types::global_dof_index> in_deal;
+    vector_space_vector_index_set.fill_index_vector(in_deal);
+    std::vector<PetscInt> in_petsc(in_deal.begin(), in_deal.end());
+
+    std::vector<types::global_dof_index> out_deal;
+    read_write_vector_index_set.fill_index_vector(out_deal);
+    std::vector<PetscInt> out_petsc(out_deal.begin(), out_deal.end());
+
+    std::vector<PetscInt> dummy;
+
+    this->do_reinit(in_petsc, dummy, out_petsc, dummy, communicator);
+  }
+
+  void
+  CommunicationPattern::reinit(
+    const std::vector<types::global_dof_index> &indices_has,
+    const std::vector<types::global_dof_index> &indices_want,
+    const MPI_Comm &                            communicator)
+  {
+    // Clean vectors from numbers::invalid_dof_index (indicating padding)
+    std::vector<PetscInt> indices_has_clean, indices_has_loc;
+    std::vector<PetscInt> indices_want_clean, indices_want_loc;
+    indices_want_clean.reserve(indices_want.size());
+    indices_want_loc.reserve(indices_want.size());
+    indices_has_clean.reserve(indices_has.size());
+    indices_has_loc.reserve(indices_has.size());
+
+    PetscInt loc         = 0;
+    bool     has_invalid = false;
+    for (const auto i : indices_has)
+      {
+        if (i != numbers::invalid_dof_index)
```

Comment:
How can this happen? An `IndexSet` should never store such an index.

Reply:
This `reinit` accepts `std::vector`, not `IndexSet`. It is used e.g. here  https://github.com/dealii/dealii/pull/14736/files/3dd5da076e273db6f1f48f24a5aad4effb9a0f0d#diff-c8636b9e39678217cab6f2d9f405735b74b68706d67096b83397caa07ff7a0e9R77

Reply:
Oh oh, of course!